src/notifier/keep_alive_monitor.py

```diff
@@ -77,18 +79,41 @@ def check_last_keep_alive(self):
                 logging.debug(f"Keep-alive check for {service.name}: Last activity {seconds_since_last} seconds ago.")
                 if seconds_since_last > self._last_keep_alive_threshold_seconds[service]:
                     message = (
-                        f"Your harvester appears to be offline! "
-                        f"No events for the past {seconds_since_last} seconds."
+                        "Your harvester appears to be offline! " f"No events for the past {seconds_since_last} seconds."
                     )
-                    logging.warning(message)
-                    events.append(
-                        Event(
-                            type=EventType.USER,
-                            priority=EventPriority.HIGH,
-                            service=EventService.HARVESTER,
-                            message=message,
-                        )
+                    new_incident = False
+                    if self._keep_alive_iteration[service] == 0:
+                        # The incident starts
+                        new_incident = True
+                        self._keep_alive_incident_time[service] = datetime.now()
+                    # Get the current notify threshold timestamp based on initial event time & iteration count.
+                    notification_threshold = exponential_backoff(
+                        incident_time=self._keep_alive_incident_time[service],
+                        interval=self._last_keep_alive_threshold_seconds[service],
+                        iteration=self._keep_alive_iteration[service],
                     )
+                    if new_incident or datetime.now() >= notification_threshold:
+                        events.append(
+                            Event(
+                                type=EventType.USER,
+                                priority=EventPriority.HIGH,
+                                service=EventService.HARVESTER,
+                                message=message,
+                                iteration=self._keep_alive_iteration[service],
+                            )
+                        )
+                        # We only increase iteration for notifications sent, not for every time the check fails.
+                        self._keep_alive_iteration[service] += 1
+                    else:
+                        message += f" The next notification won't be sent before {notification_threshold}"
+                    # Message is logged regardless of threshold
+                    logging.warning(message)
+                else:
+                    # All is fine, reset iteration
+                    if self._keep_alive_iteration[service] > 0:
+                        logging.info(f"incident for {service} is over")
+                    self._keep_alive_iteration[service] = 0
+                    self._keep_alive_incident_time[service] = None
```

Comment:
Now that I'm looking at this, it's actually making this class more complex than I'd like (and really, a pain to maintain). It would be justified to refactor this into its own class. This will allow us to cover the functionality by unit tests too. I'd be hesitant to merge before we have some simple unit-test coverage.

As I see it, It should keep state for a single service, and then we can create instances for every service here. I'd propose the following interface (name can be better):

```
class EventThrottleHelper:
    def __init__(self, interval_seconds: int):
        self.event_counter = 0
        self.first_event_time = None
        self.interval_seconds = interval_seconds

    def should_trigger_event(self):
        time_now = datetime.now()

        # Auto-detect when to reset event_counter and
        # first_event_time based on current time,
        # interval_seconds and first_event_time
```

where you initialize the class with the `self._last_keep_alive_threshold_seconds[service]` and everything else can be encapsulated.

The change in `KeepAliveMonitor` would then be very minimal, just extend the if-statement on line 80:

```
if seconds_since_last > self._last_keep_alive_threshold_seconds[service] and \
    throttle_helper[service].should_trigger_event():
```